### PR TITLE
Add Tekton Design Principles

### PR DESCRIPTION
**Pull request by jerop**
The goal of these principles is to provide a basis for design choices and help in making tradeoffs. The design principles are reusability, simplicity, flexibility and conformance. 

The principles were initially proposed and discussed in [this google doc](https://docs.google.com/document/d/1_iqwslxAOx-SPYosnTFS5kOZcmxcBWpuIULZHQHP85Q/edit?usp=sharing) that's visible to members of tekton-dev@. Thank you for the contributions!

Related issue: [Add contributor and reviewer expectations](https://github.com/tektoncd/community/pull/133).

### CODE REVIEW

File: design-principles.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Tekton Design Principles
+
+## Reusability
+- Existing features should be reused when possible instead of adding new ones.
```
Comment (vdemeester) [inline design-principles.md:4]: Another way to say this (or a complement to this) would be : when implemented a feature, make sure you tried with existing features so that you can make sure you have a clear reason to add this feature.

File: design-principles.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+## Simplicity 
+- Tekton should contain only the bare minimum and simplest features needed to meet the largest number of CI/CD use cases.
+- Prefer a [simple](https://www.infoq.com/presentations/Simple-Made-Easy/) solution that solves most use cases to a complex solution that solves all use cases (can be revisited later).
+- New features should be consistent with existing components, in structure and behavior, to make learnability, trialability and adoption easy.
+- Any new feature should have been previously discussed and agreed upon in a [Tekton Enhancement Proposal](https://github.com/tektoncd/community/tree/master/teps). 
+- In design docs, demonstrate that the proposed feature is absolutely necessary. What’s the current experience without the feature and how challenging is it?
+
+## Flexibility
+- Tekton has a ton of flexibility, which means a lot of things can be implemented by some kind of plugin. When considering adding something to Tekton itself, we should consider and exhaust all opportunities to implement it using one of the existing plugin mechanisms.
```
Comment (vdemeester) [inline design-principles.md:16]: I am not sure I fully *understand* the word `plugin` in that context.

File: design-principles.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+- Specific choices (kubernetes, github) should be implemented at the step or DSL levels by the users, not at the Task or Pipeline levels to keep Tekton flexible.
```
Comment (vdemeester) [inline design-principles.md:17]: I am not sure I completely follow. I think this needs an example or something — I understand what we are getting at but it's not that clear to me when reading

File: design-principles.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+- Any new feature should have been previously discussed and agreed upon in a [Tekton Enhancement Proposal](https://github.com/tektoncd/community/tree/master/teps). 
+- In design docs, demonstrate that the proposed feature is absolutely necessary. What’s the current experience without the feature and how challenging is it?
+
+## Flexibility
+- Tekton has a ton of flexibility, which means a lot of things can be implemented by some kind of plugin. When considering adding something to Tekton itself, we should consider and exhaust all opportunities to implement it using one of the existing plugin mechanisms.
+- Specific choices (kubernetes, github) should be implemented at the step or DSL levels by the users, not at the Task or Pipeline levels to keep Tekton flexible. 
+- When a specific choice (tool, resource, language, etc) has to be made at the Task or Pipeline levels, users should be able to extend it to add their own choices.
+- When a specific choice is in consideration, evaluate what we’re coupling Tekton to and what it means in terms of support and maintenance.
+- Avoid implementing templating logic; prefer variable replacement.
```
Comment (vdemeester) [inline design-principles.md:20]: Isn't this highly specific ? (on "coupled" to `tektoncd/pipeline`)

Comment (dibyom) [reply design-principles.md:20]: Applies to triggers too 😛 

I think it might be useful to link to the design doc on whether to template or not which provides the basis for our recommendation!

File: design-principles.md (hunk 2 above)
Comment (jerop) [reply design-principles.md:16]: I understood this to mean that implementations can be made in the step level or DSL level and that people can now use custom tasks -- it was from @ImJasonH so maybe he can clarify further

File: design-principles.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+- Avoid implementing our own expression syntax; when required prefer existing languages which are widely used and include supporting development tools.
+
+## Conformance
+- Tekton features should work as the user expects in varied environment setup.
+- Tekton should not contain kubernetes-specific features in the API as much as possible. When kubernetes-specific features have to be added, they should be explicitly called out in the design docs and consider shunting them together into a section of the API, such as podTemplate.
```
[the remark this answers is not in the document]
Comment (jerop) [reply design-principles.md:25]: I understand kubernetes-specific here means things such as configurations for a Pod

/cc @ImJasonH

File: design-principles.md (hunk 3 above)
Comment (jerop) [reply design-principles.md]: added an example that Tekton API should not be aware of git concepts, such as branches

File: design-principles.md (hunk 4 above)
Comment (jerop) [reply design-principles.md]: good catch, linked the doc :)

**Comment by bobcatfish** [comment]
I'm very excited! Any thoughts I had were addressed in the original doc so:

/lgtm

Waiting for review from @vdemeester @afrittoli  (and anyone/everyone else who would like to chime in of course) before merging

**Review by vdemeester — COMMENTED** [review]
/lgtm

File: design-principles.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+- New features should be consistent with existing components, in structure and behavior, to make learnability, trialability and adoption easy.
+- Any new feature should have been previously discussed and agreed upon in a [Tekton Enhancement Proposal](https://github.com/tektoncd/community/tree/master/teps). 
+- In design docs, demonstrate that the proposed feature is absolutely necessary. What’s the current experience without the feature and how challenging is it?
+
+## Flexibility
+- Tekton has a ton of flexibility, which means a lot of things can be implemented by some kind of plugin, such as using `CustomTasks` or in the `Step` level. When considering adding something to Tekton itself, we should consider and exhaust all opportunities to implement it using one of the existing plugin mechanisms.
+- Specific choices should be implemented at the step or DSL levels by the users, not at the Task or Pipeline levels to keep Tekton flexible. Such as Tekton should not be aware of git concepts, such as branches.
```
Comment (afrittoli) [inline design-principles.md:17]: I agree with this, but the step vs task/pipeline distinction is not 100% clear to me.
Does this mean for example that we want a `CloneRepo` task that includes branch logic in the step, `if git then elif snv then ...`? 
I think having tasks specifically designed and tested for each SCM, selected with conditionals (when expressions?_) etc would be a valid alternative.

Comment (bobcatfish) [reply design-principles.md:17]: that makes sense to me, maybe we can reword this as something like:

- To keep Tekton flexible, specific choices should be implemented within Tasks or at the DSL level by users, not at the level of features provided by Tekton in the Task or Pipeline APIs. Such as Tekton should not be aware of git concepts, such as branches.

File: design-principles.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+- When a specific choice (tool, resource, language, etc) has to be made at the Task or Pipeline levels, users should be able to extend it to add their own choices.
+- When a specific choice is in consideration, evaluate what we’re coupling Tekton to and what it means in terms of support and maintenance.
+- [Avoid implementing templating logic](https://docs.google.com/document/d/1h_3vSApIsuiwGkrqSiegi4NVaYG4oVzBquGAhIN6qGM/edit#heading=h.6kxvcvm7rs3r); prefer variable replacement.
+- Avoid implementing our own expression syntax; when required prefer existing languages which are widely used and include supporting development tools.
+
+## Conformance
+- Tekton features should work as the user expects in varied environment setup.
+- Tekton should not contain kubernetes-specific features, such as configuations for a `Pod`, in the API as much as possible. When kubernetes-specific features have to be added, they should be explicitly called out in the design docs and consider shunting them together into a section of the API, such as `podTemplate`.  
+- In design docs, discuss how the proposal affects conformance.
```
Comment (afrittoli) [inline design-principles.md:26]: NIT: it would be good to have a link here to what we mean by conformance - Jason's PR with the API spec is not merged yet thought, so we could add that later

File: design-principles.md (hunk 6 above)
Comment (afrittoli) [reply design-principles.md]: @jerop thanks for the update. 
Sorry to insist on this, but I feel it would be better to be specific about the Tekton API, as @bobcatfish suggested. 
Tekton should avoid being opinionated in the Task and Pipeline *API* , and Tasks (e.g. from the catalog) should be a valid place where to be specific and opinionated.

Comment (jerop) [reply design-principles.md]: thanks for making that point clearer, updated it: https://github.com/tektoncd/community/pull/171/files#diff-0899720b7e7bf227aaa00d838f2c9e76R17

File: design-principles.md (hunk 7 above)
Comment (jerop) [reply design-principles.md]: agreed, added a link to the API spec tep: https://github.com/tektoncd/community/blob/master/teps/0012-api-spec.md

**Review by vdemeester — APPROVED** [review]
/cc @afrittoli

**Comment by bobcatfish** [comment]
Hey @jereop, just realized we didn't add any links to this - can we add a link from the main README?
